### PR TITLE
Add support for RISC-V semihosting

### PR DESCRIPTION
**Pull request by arichardson**
See https://github.com/riscv-software-src/riscv-semihosting for the current specification. Almost all code is shared with the Arm implementation.

Tested by running some binaries built with [picolibc](https://github.com/picolibc/picolibc).

### CODE REVIEW

**Comment by powerjg** [comment]
Wow! This is a big contribution!

We're (the maintainers) currently focused on the v23.1 release, but after the holidays we'll definitely help out with getting this merged! You may want to join the developer meeting tomorrow (#661), if that's possible for you, to discuss this and other RISC-V improvements.

File: src/arch/arm/ArmSemihosting.py
Comment (rogerchang23424) [inline src/arch/arm/ArmSemihosting.py:39]: Please remove the parameters that are specified in the `BaseSemihosting`

Comment (arichardson) [reply src/arch/arm/ArmSemihosting.py:39]: Good catch, will update

**Comment by rogerchang23424** [comment]
Another

**Comment by rogerchang23424** [comment]
The debug flag `Semihosting` is defined in the `src/arch/arm`. It need to move to `src/arch/generic`.

https://github.com/gem5/gem5/blob/c66862f6e3fbad5b0066c847aea8c5aef62b367e/src/arch/arm/SConscript#L135

**Comment by andysan** [comment]
@giactra: Could you have a look at this? It looks like quite a lot of Arm specific stuff has ended up in the generic file.

**Comment by arichardson** [comment]
> @giactra: Could you have a look at this? It looks like quite a lot of Arm specific stuff has ended up in the generic file.

Could you give me an example? RISC-V semihosting reuses the same semihosting ops as ARM (https://github.com/ARM-software/abi-aa/blob/2023Q3/semihosting/semihosting.rst#semihosting-operations) so I believe all of this needs to be in the generic file?

File: src/arch/generic/SConscript
Comment (rogerchang23424) [inline src/arch/generic/SConscript]: The tags `arm isa` required to be removed, the flag will not build if the `USE_ARM_ISA` is not set.

Comment (arichardson) [reply src/arch/generic/SConscript]: Oh yes stupid copy paste error. Will fix tomorrow

File: configs/example/riscv/fs_linux.py
Comment (rogerchang23424) [inline configs/example/riscv/fs_linux.py]: The parameter is `NULL` to make it happy.

File: src/arch/riscv/semihosting.cc
Comment (rogerchang23424) [inline src/arch/riscv/semihosting.cc]: The spec doesn't required the sequence of instruction to be in the same instruction block(16 bytes), right?

**Comment by andysan** [comment]
> > @giactra: Could you have a look at this? It looks like quite a lot of Arm specific stuff has ended up in the generic file.
> 
> Could you give me an example? RISC-V semihosting reuses the same semihosting ops as ARM (https://github.com/ARM-software/abi-aa/blob/2023Q3/semihosting/semihosting.rst#semihosting-operations) so I believe all of this needs to be in the generic file?

The feature bytes and the exit reasons both seem somewhat Arm specific. I would imagine that there will be a point where the RISC-V implementation will need to modify these and add RV-specific features. The fact that the function tables are architecture specific, also implies that the feature structure needs to be architecture specific.

**Comment by giactra** [comment]
> Could you give me an example? RISC-V semihosting reuses the same semihosting ops as ARM (https://github.com/ARM-software/abi-aa/blob/2023Q3/semihosting/semihosting.rst#semihosting-operations) so I believe all of this needs to be in the generic file?

Hey Alex, I tried to have a look at the link you mentioned in the first PR commit [1] and while I see the specification of the intruction trap sequence and register definition, I don't see any reference to the semihosting ops nor I find a line that describes them as being the same as Arm. Could you point me to the relevant documentation?

Thanks!

[1]: https://github.com/riscv-software-src/riscv-semihosting/blob/main/riscv-semihosting-spec.adoc#register

File: src/arch/generic/semihosting.hh
Comment (andysan) [inline src/arch/generic/semihosting.hh]: To avoid confusion, could you add a link to the normative specification (https://github.com/ARM-software/abi-aa/blob/main/semihosting/semihosting.rst) here? There used to be a link in the R-V specification, but it was dropped for some reason. It's a bit weird that R-V implements some of the quirks we needed for backwards compatibility and pre-Armv8 features, so I think it might be worth clarifying that it is literally implementing the Arm spec.

File: src/arch/riscv/semihosting.cc
Comment (arichardson) [reply src/arch/riscv/semihosting.cc]: Oh indeed that is too strict, they just need to be on the same page (if paging is enabled).

> If paging is in use in the current mode, this sequence must not cross a page boundary as the semihosting system must be able to check for the semihosting sequence without needing data from potentially missing pages.

File: src/arch/generic/semihosting.hh
Comment (arichardson) [reply src/arch/generic/semihosting.hh]: I believe the intention behind the RISC-V semihosting spec was to be able to reuse as much as possible from existing ARM implementations (both emulators as well as libc support). It's not a final spec, so I believe it's still possible to make some changes to it to remove old quirks (although QEMU has implemented the current version since 6.0 so not sure if there is much that can be changed at this point).

CC: @keith-packard

File: configs/example/riscv/fs_linux.py
Comment (arichardson) [reply configs/example/riscv/fs_linux.py]: I still need to initialize this variable so I can pass it to the constructor further down.

I guess the alternative would be to use something like

```
workload_args = dict()
if args.semihosting:
    workload_args["semihosting"] = RiscvSemihosting(
        files_root_dir=args.semihosting_root,
        cmd_line=args.kernel,
    )
if args.bare_metal:
    system.workload = RiscvBareMetal(**workload_args)
    system.workload.bootloader = args.kernel
else:
    system.workload = RiscvLinux(**workload_args)
    system.workload.object_file = args.kernel
```

but I'm not convinced that is cleaner.

File: src/arch/generic/semihosting.hh
Comment (keith-packard) [reply src/arch/generic/semihosting.hh]: Yup, the goal was to share the same spec and implementations to avoid fragmentation. I just inherited the bits from the actual authors (mostly Palmer and Megan, I believe?) and got them published so I could upstream the qemu support and use that for picolibc testing.

File: configs/example/riscv/fs_linux.py
Comment (rogerchang23424) [reply configs/example/riscv/fs_linux.py:181]: > I still need to initialize this variable so I can pass it to the constructor further down.
> 
> I guess the alternative would be to use something like
> 
> ```
> workload_args = dict()
> if args.semihosting:
>     workload_args["semihosting"] = RiscvSemihosting(
>         files_root_dir=args.semihosting_root,
>         cmd_line=args.kernel,
>     )
> if args.bare_metal:
>     system.workload = RiscvBareMetal(**workload_args)
>     system.workload.bootloader = args.kernel
> else:
>     system.workload = RiscvLinux(**workload_args)
>     system.workload.object_file = args.kernel
> ```
> 
> but I'm not convinced that is cleaner.

LGTM, works in my side.

File: src/arch/riscv/semihosting.cc
Comment (rogerchang23424) [reply src/arch/riscv/semihosting.cc]: Ok, I would like to check if the instruction sequence is in the same 4K Bytes page block. It will reduce the strict checking.

**Comment by arichardson** [comment]
Add the latest updates as fixup commits to the end - will rebase to create a sensible history once reviewed.

File: configs/example/riscv/fs_linux.py
Comment (rogerchang23424) [reply configs/example/riscv/fs_linux.py]: > ```
> workload_args = dict()
> if args.semihosting:
>     workload_args["semihosting"] = RiscvSemihosting(
>         files_root_dir=args.semihosting_root,
>         cmd_line=args.kernel,
>     )
> if args.bare_metal:
>     system.workload = RiscvBareMetal(**workload_args)
>     system.workload.bootloader = args.kernel
> else:
>     system.workload = RiscvLinux(**workload_args)
>     system.workload.object_file = args.kernel
> ```

Cloud you apply the code?

Comment (rogerchang23424) [reply configs/example/riscv/fs_linux.py]: bump

**Comment by powerjg** [comment]
@andysan and @giactra, do you see any changes that are needed before this is merged?

File: src/arch/generic/BaseSemihosting.py
Comment (giactra) [inline src/arch/generic/BaseSemihosting.py]: This code is basically copied/extracted from ArmSemihosting.py so it should bear the original copyright holder (Arm Limited), whereas here you replaced Arm with Riscv

**Review by giactra — CHANGES_REQUESTED** [review]
Thanks @powerjg I believe there is still an unaddressed request from Andreas to include a reference to the semihosting documentation https://github.com/gem5/gem5/pull/681/files#r1427809702, also I spotted a copyright issue (see below in review).

Otherwise LGTM, though I am running some tests to make sure nothing is broken... We should probably add some semihosting unit tests to our regressions in the long term

**Comment by ivanaamit** [comment]
Hi @arichardson, as it has been three weeks without updates on this PR, I wanted to check in and see if you are currently working on incorporating the requested changes. Thank you.

**Comment by arichardson** [comment]
> Hi @arichardson, as it has been three weeks without updates on this PR, I wanted to check in and see if you are currently working on incorporating the requested changes. Thank you.

Apologies, I missed these comments in the flood of other GitHub notifications. Will rebase and update shortly

**Comment by ivanaamit** [comment]
Hi @arichardson, another friendly reminder to check if you are working on updating this PR. Thank you.

**Comment by arichardson** [comment]
> Hi @arichardson, another friendly reminder to check if you are working on updating this PR. Thank you.

I am planning to get to this today or tomorrow. Thank you for the reminder!

**Comment by arichardson** [comment]
Sorry for the delay, I've now rebased and added two new fixup commits that should address the outstanding comments. Please let me know if this looks good and I'll squash the commits to a sensible history.

File: src/arch/generic/BaseSemihosting.py
Comment (arichardson) [reply src/arch/generic/BaseSemihosting.py]: I failed to exclude this line from my search+replace :( Should be fixed now.

**Review by rogerchang23424 — APPROVED** [review]
LGTM on RISC-V part, Thanks

**Comment by giactra** [comment]
Thanks @arichardson for updating the header! Though there is still the remaining request from Andreas to be addressed  https://github.com/gem5/gem5/pull/681/files#r1427809702 and then this is probably good to go

**Comment by arichardson** [comment]
Sorry for the delay. Updated the PR to add pointers to the documentation and rebased and edited the history into two commits.

**Review by giactra — DISMISSED** [review]
Thanks for the patience, looks good from my side, waiting for @andysan to provide his feedback as well

File: src/arch/generic/BaseSemihosting.py
Comment (andysan) [inline src/arch/generic/BaseSemihosting.py]: This should presumably have been in the previous commit.

File: src/arch/generic/BaseSemihosting.py
Comment (andysan) [reply src/arch/generic/BaseSemihosting.py]: It seems like this was fixed in the wrong commit.

File: src/sim/workload.hh
Comment (andysan) [inline src/sim/workload.hh:107]: I'd suggest not adding this here unless it's used by other architectures.

**Review by andysan — DISMISSED** [review]

**Comment by ivanaamit** [comment]
@arichardson, could you please address Andreas's comments? Thank you.

**Comment by arichardson** [comment]
> @arichardson, could you please address Andreas's comments? Thank you.

Sorry I missed these since I get way too many GitHub notifications. Will try to address ASAP

**Comment by ivanaamit** [comment]
@arichardson, we would appreciate it if you could address the pending comments. Additionally, you should resolve conflicting files for this PR to be merged. Thank you.

**Comment by ivanaamit** [comment]
@arichardson, it would be great if you could simply rebase, and then we can merge this PR. Andreas's comments are not blocking the merge; you can address them if you have time. Thank you.

File: src/sim/workload.hh
Comment (rogerchang23424) [reply src/sim/workload.hh:107]: We get the semihost from workload `src/arch/riscv/insts/standard.cc`. https://github.com/gem5/gem5/pull/681/commits/e48114c6581d9493030b155717e4d93ec1600957#diff-e0c1e51647bad6c3927cecd92b966ef5f523489faaf39bc4647b1b712a0e39b8

There are three types of workloads used in RISC-V system, `RiscvBareMetal`, `RiscvLinux` and `RiscvBootloaderKernelWorkload`. I think adding the getSimihosting in generic `Workload` is make sense to me

**Comment by BobbyRBruce** [comment]
I've resolved the merge conflicts via a merge commit from develop to this branch. Given @andysan's comments about things being in the wrong commits anyway, I'll submit this as a squash-merge once the tests pass.

**Comment by arichardson** [comment]
Sorry for the delay here and thanks for taking over!

File: src/arch/generic/semihosting.hh
Comment (arichardson) [inline src/arch/generic/semihosting.hh]:
```suggestion
 * Note: The RISC-V semihosting specification reuses the Arm semihosting
 * interfaces (https://github.com/riscv-non-isa/riscv-semihosting).
```

Comment (arichardson) [reply src/arch/generic/semihosting.hh]: This has now been fixed in the latest RISC-V semihosting spec.
There is a 0.3 release that says the following:

> The RISC-V semihosting specification adapts the design of publicly available and open source
> semihosting specification [1] to minimize the development effort. Only the mechanism of invoke a
> semihosting service (aka semihosting binary interface) is architecture specific whereas all defined
> semihosting services [1] are architecture independent and resemble system calls of an unix-style
> operating system.

With [1] pointing to the ARM semihosting spec.

File: src/arch/riscv/RiscvFsWorkload.py
Comment (rogerchang23424) [inline src/arch/riscv/RiscvFsWorkload.py]: This part should not be changed in the PR
```suggestion
        True,
        "Use bootloader entry point as reset vector. "
        "If `auto_reset_vect` is true, the `reset_vect` parameter is ignored.",
```

**Review by powerjg — APPROVED** [review]

**Review by BobbyRBruce — APPROVED** [review]

**Pull request merged**
